Documentation: Document failure domains of dex for tectonic.

rithujohn191 opened the pull request.

robszumski commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,77 @@
+
+### Dex Instance(s) Fails or Becomes Unreachable
+
+This scenarios includes failures of the load balancer fronting Dex and of the backend instances of Dex behind that load balancer. Generally, Dex is run as a deployment on top of Kubernetes; meaning if a Dex instance stops responding to the /identity/healthz endpoint it will be rescheduled.
```
`scenarios` (no s on the end)

robszumski replied:
is this really an "and" or is it "or"?

does this present the same to the end user/admin either way?

robszumski replied:
maybe mention that we run multiple

robszumski commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,77 @@
+
+**Downtime** of all Dex API endpoints will incur if ALL instances of Dex are down OR the load balancer is down/unreachable by clients.
+   * Low risk to clients that cache keys, like the Kubernetes API server, from seconds to minutes of downtime.
+   * Risk to clients that require access to Dex APIs to work. e.g. Starting a Kubernetes API server.
```
to keep the progression going, would this be `Higher risk to clients...`

robszumski commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,77 @@
+**Downtime** of all Dex API endpoints will incur if ALL instances of Dex are down OR the load balancer is down/unreachable by clients.
+   * Low risk to clients that cache keys, like the Kubernetes API server, from seconds to minutes of downtime.
+   * Risk to clients that require access to Dex APIs to work. e.g. Starting a Kubernetes API server.
+   * Risk if Dex doesn't recover that keys won't be rotated before expiration. This would require downtime of hours.
```
in line with above, `Highest risk if ...`

robszumski commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,77 @@
+
+**Mean time to recovery (MTTR)** for Dex from an authentication provider being down is 0 seconds as long as the user has logged in. Authentication providers are only required when a user attempts to login.
+
+**Downtime** of all Dex API endpoints will incur IF all instances of Dex are down.
```
found this confusing. is it basically saying that "if dex is down, dex is down"?

ericchiang replied:
Maybe "dex will be unable to authenticate new users or refresh existing credentials"?

robszumski commented on Documentation/admin/failure-domains-dex.md:
````diff
@@ -0,0 +1,77 @@
+
+The easiest ways for testing outage is to scale the Dex deployment on Kubernetes down to 0. For example for clusters installed using the the Tectonic Installer simply run:
+```
+kubectl scale deployment tectonic-identity -n tectonic-system --replicas=0
````
add `$`, remove newline below

robszumski commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,77 @@
+
+## FAQ
+Q: How does a user recover or reconfigure a cluster in case Dex fails?
+A: If Dex is down and the user is unable to login via the Tectonic console they can make use of the kubeconfig generated by the installer which can be found in the [assets folder](https://coreos.com/tectonic/docs/latest/admin/assets-zip.html). This will allow the user to access the kubernetes API directly.
```
on github you can link to the relative path

ericchiang commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,77 @@
+* **Tectonic Console**: It depends on Dex to login users and create a browser session cookie via the graphical login interface.
+* **Kubectl**: Uses an identity token to authenticate to the API server and every few hours uses a refresh token to renew it.
+               Known issue: kubectl makes a roundtrip on every request, this will be fixed in kubectl v1.7.0+ and potentially a patch release to v1.6. Issue documented [here][kubernetes-issue]
+* **API server**: It uses all of the Dex endpoints, including periodic refreshes from the `/identity/keys` endpoint, to authenticate users making requests to the API server. The API server does not require access to Dex on every request and caches these public keys allowing the API to be resilient to temporary outages of Dex, the load balancer, or the network.
```
API server actually only talks to /.well-known and /keys. e.g. if the token or auth endpoints fail the API server will be fine.

ericchiang commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,77 @@
+
+This scenarios includes failures of the load balancer fronting Dex and of the backend instances of Dex behind that load balancer. Generally, Dex is run as a deployment on top of Kubernetes; meaning if a Dex instance stops responding to the /identity/healthz endpoint it will be rescheduled.
+
+**Mean time to recovery (MTTR)** for a Dex instance that has failed will be 5s-60s to launch a new container or download the Dex container and launch.
```
We also might want to note that we should increase the number of dex instances deployed. That would mean we could aim for zero downtime from misbehaving dex instances (not storage failures).

Maybe a "future improvements" section?

rithujohn191 commented:
Thank you for the feedback. Addressed all the suggestions. If you have an opinion of if we should call it dex or tectonic identity I can make changes accordingly. I am ok with both

sym3tri commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,76 @@
+
+## FAQ
+Q: How does a user recover or reconfigure a cluster in case Dex fails?
+A: If Dex is down and the user is unable to login via the Tectonic console they can make use of the kubeconfig generated by the installer which can be found in the [assets folder](assets-zip.html). This will allow the user to access the kubernetes API directly.
```
Nit: the answer part does not render on a new line.

sym3tri commented on Documentation/admin/failure-domains-dex.md:
```diff
@@ -0,0 +1,76 @@
+A: If Dex is down and the user is unable to login via the Tectonic console they can make use of the kubeconfig generated by the installer which can be found in the [assets folder](assets-zip.html). This will allow the user to access the kubernetes API directly.
+
+Q: Is there a circular dependency between the API server and Dex? For example Dex is used to authenticate requests and Dex also uses the Kubernetes API server to persist data.
+A: The Kubernetes API server can authenticate requests in three ways: using OIDC via Dex for users, using TLS client certificate for disaster recovery by users, and [service accounts](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/) for on-cluster services. The Dex service uses a service account to talk to the API server which resolves the dependency.
```
same

zbwright commented:
my 2cents: it may be a good idea to call it Tectonic Identity, which users will associate with CoreOS, rather than something with no brand recognition.

rithujohn191 commented:
Replaced all occurrences of dex with Tectonic Identity

philips commented:
@rithujohn191 the diagram still says dex.

rithujohn191 commented:
Changed the diagram to say Tectonic Identity instead of Dex

sym3tri commented:
LGTM any subsequent changes/additions can be made in new PRs.
